src/connection_string.ts

```diff
@@ -785,26 +775,792 @@ function parseURI(uri: string): { srv: boolean; url: URL; hosts: string[] } {
+
+  if (srv && hostList.length !== 1) {
+    throw new MongoParseError('mongodb+srv URI cannot have multiple service names');
+  }
+  if (srv && hostList[0].includes(':')) {
+    throw new MongoParseError('mongodb+srv URI cannot have port number');
+  }
+
   return {
-    srv: protocol.includes('srv'),
-    url: new URL(`${protocol.toLowerCase()}://${authString}@dummyHostname${rest}`),
+    srv,
+    url,
     hosts: hosts.split(',')
   };
 }
 
+const TRUTHS = new Set(['true', 't', '1', 'y', 'yes']);
+const FALSEHOODS = new Set(['false', 'f', '0', 'n', 'no', '-1']);
+function getBoolean(name: string, value: unknown): boolean {
+  if (typeof value === 'boolean') return value;
+  const valueString = String(value).toLowerCase();
+  if (TRUTHS.has(valueString)) return true;
+  if (FALSEHOODS.has(valueString)) return false;
+  throw new TypeError(`For ${name} Expected stringified boolean value, got: ${value}`);
+}
+
+function getInt(name: string, value: unknown): number {
+  if (typeof value === 'number') return Math.trunc(value);
+  const parsedValue = Number.parseInt(String(value), 10);
+  if (!Number.isNaN(parsedValue)) return parsedValue;
+  throw new TypeError(`Expected ${name} to be stringified int value, got: ${value}`);
+}
+
+function getUint(name: string, value: unknown): number {
+  const parsedValue = getInt(name, value);
+  if (parsedValue < 0) {
+    throw new TypeError(`${name} can only be a positive int value, got: ${value}`);
+  }
+  return parsedValue;
+}
+
+function toRecord(value: string): Record<string, any> {
+  const record = Object.create(null);
+  const keyValuePairs = value.split(',');
+  for (const keyValue of keyValuePairs) {
+    const [key, value] = keyValue.split(':');
+    record[key] = value;
+  }
+  return record;
+}
+
+const DEFAULT_PK_FACTORY = {
+  createPk(): ObjectId {
+    // We prefer not to rely on ObjectId having a createPk method
+    return new ObjectId();
+  }
+};
+
+const defaultOptions = new Map<string, unknown>(
+  ([
+    ['readPreference', ReadPreference.primary],
+    ['autoEncryption', {}],
+    ['compression', 'none'],
+    ['compressors', 'none'],
+    ['connectTimeoutMS', 30000],
+    ['dbName', 'test'],
+    ['directConnection', false],
+    ['driverInfo', {}],
+    ['forceServerObjectId', false],
+    ['heartbeatFrequencyMS', 10000],
+    ['keepAlive', true],
+    ['keepAliveInitialDelay', 120000],
+    ['localThresholdMS', 0],
+    ['logger', new Logger('MongoClient')],
+    ['maxIdleTimeMS', 0],
+    ['maxPoolSize', 100],
+    ['minPoolSize', 0],
+    ['monitorCommands', false],
+    ['noDelay', true],
+    ['numberOfRetries', 5],
+    ['pkFactory', DEFAULT_PK_FACTORY],
+    ['raw', false],
+    ['retryReads', true],
+    ['retryWrites', true],
+    ['serverSelectionTimeoutMS', 30000],
+    ['serverSelectionTryOnce', false],
+    ['socketTimeoutMS', 0],
+    ['waitQueueTimeoutMS', 0],
+    ['zlibCompressionLevel', 0]
+  ] as [string, any][]).map(([k, v]) => [k.toLowerCase(), v])
```

Comment:
I see that `toLowerCase` is being called in at least three places, not to mention when you do `setOption` you are doing a case-insensitive regex match. Is it possible that we're doing things a bit backwards here? 

You already have a registry of every known option name, in the preferred camelCase form that we want to end up with. Why not iterate through the keys of that registry, check if the property exists (using a case-insensitive match) in the uri options or programmatically provided options (determine precdence here), and then operate on those values in place? I think that approach wouldn't require any clever case translation outside of the case-insensitive string match, not to mention I think it would require less algorithmic complexity on the whole.

Reply:
I've gathered all the case fixing into a CaseInsensitiveMap implementation, makes this a lot less repetitive, I did move the loop to go over the registry however my thinking behind using allOptions (all the specified + defaults) is just that it bounds the loop to a smaller number of iterations but it is a small difference in the grand scheme.